### PR TITLE
Add basic handling for generic params variance

### PR DESCRIPTION
**Pull request by BarkingBad**

### CODE REVIEW

File: plugins/javadoc/src/main/kotlin/org/jetbrains/dokka/javadoc/signatures/JavadocSignatureProvider.kt [hunk 1]
```diff
@@ -150,7 +150,7 @@ class JavadocSignatureProvider(ctcc: CommentsToContentConverter, logger: DokkaLo
                 annotationsBlock(t)
             }
             signatureWithoutModifiers {
-                text(t.name)
+                text(t.name) // Investigate if java classes can be somehow variant
```
Comment (BarkingBad) [inline plugins/javadoc/src/main/kotlin/org/jetbrains/dokka/javadoc/signatures/JavadocSignatureProvider.kt:153]: Then it should be something like that `? extends T extends SomeInterface`

Comment (BarkingBad) [reply plugins/javadoc/src/main/kotlin/org/jetbrains/dokka/javadoc/signatures/JavadocSignatureProvider.kt:153]: On the other hand we should somehow indicate that when the class comes from Kotlin sources

File: plugins/kotlin-as-java/src/main/kotlin/signatures/JavaSignatureProvider.kt [hunk 2]
```diff
@@ -141,7 +141,7 @@ class JavaSignatureProvider(ctcc: CommentsToContentConverter, logger: DokkaLogge
     private fun signature(t: DTypeParameter) =
         t.sourceSets.map {
             contentBuilder.contentFor(t, styles = t.stylesIfDeprecated(it), sourceSets = setOf(it)) {
-                text(t.name.substringAfterLast("."))
+                text(t.name.substringAfterLast(".")) // Investigate if java classes can be somehow variant
```
Comment (BarkingBad) [inline plugins/kotlin-as-java/src/main/kotlin/signatures/JavaSignatureProvider.kt:144]: Same as above

File: plugins/base/src/test/kotlin/signatures/SignatureUtils.kt [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+import org.jsoup.Jsoup
+import org.jsoup.nodes.Element
+import utils.TestOutputWriter
+
+fun TestOutputWriter.renderedContent(path: String = "root/example.html") =
+    contents.getValue(path).let { Jsoup.parse(it) }.select("#content")
+        .single()
+
+fun Element.signature() = select("div.symbol.monospace")
+fun Element.firstSignature() = signature().first()
```
Comment (sellmair) [inline plugins/base/src/test/kotlin/signatures/SignatureUtils.kt:12]: Hey small tip: 
You might want to enable the intelliJ setting: 
"New line on save" to ensure that you always end files with a newline ☺️

File: core/src/main/kotlin/model/Documentable.kt [hunk 4]
```diff
@@ -366,6 +389,12 @@ enum class FunctionModifiers {
     NONE, FUNCTION, EXTENSION
 }
 
+fun Variance<TypeParameter>.withNewDri(dri: DRI) = when(this) {
```
Comment (sellmair) [inline core/src/main/kotlin/model/Documentable.kt:392]: NIT: I think you can easily drop the `new` part. `withDRI` or `withDri` will be perfectly understandable IMHO

Comment (sellmair) [inline core/src/main/kotlin/model/Documentable.kt:392]: NIT: I think you can easily drop the new part. withDRI or withDri will be perfectly understandable IMHO

File: plugins/javadoc/src/main/kotlin/org/jetbrains/dokka/javadoc/signatures/JavadocSignatureProvider.kt (hunk 1 above)
Comment (sellmair) [reply plugins/javadoc/src/main/kotlin/org/jetbrains/dokka/javadoc/signatures/JavadocSignatureProvider.kt:153]: I think `? extends SomeInterface` or `? super SomeInterface`  or am I wrong?  🤔

**Review by sellmair — COMMENTED** [review]
Very, nice ☺️ 👍

File: plugins/javadoc/src/main/kotlin/org/jetbrains/dokka/javadoc/signatures/JavadocSignatureProvider.kt (hunk 1 above)
Comment (BarkingBad) [reply plugins/javadoc/src/main/kotlin/org/jetbrains/dokka/javadoc/signatures/JavadocSignatureProvider.kt:153]: Yeah, but that works for properties/parameters. There is no such mechanism for class declaration, which we here discuss.

Comment (BarkingBad) [reply plugins/javadoc/src/main/kotlin/org/jetbrains/dokka/javadoc/signatures/JavadocSignatureProvider.kt:153]: I mean, in Kotlin we can have `class Klass<out T : SomeInterface>`. To accomplish that in Java, you should have something like `class Klass<? extends T extends SomeInterface>()` which is of course not a correct syntax.

File: plugins/base/src/test/kotlin/signatures/SignatureUtils.kt (hunk 3 above)
Comment (BarkingBad) [reply plugins/base/src/test/kotlin/signatures/SignatureUtils.kt:12]: Ok, will change, thanks

File: core/src/main/kotlin/model/Documentable.kt (hunk 4 above)
Comment (BarkingBad) [reply core/src/main/kotlin/model/Documentable.kt:392]: Will change

**Pull request merged**